File: euler-ode/bin/dpm/plot-orbit.py
import os, json, yaml
import numpy as np
import matplotlib.pyplot as plt

# Problem invariant
def energy(q, p):
    theta_1, theta_2 = q
    p_1, p_2 = p
    cdiff = np.cos(theta_1 - theta_2)
    return 0.5*(p_1**2 + 2*p_2**2 - 2*p_1*p_2*cdiff)/(2 - cdiff**2) + 3 - 2*np.cos(theta_1) - np.cos(theta_2)

# Configuration
with open('conf/config.json', 'r') as cli_cfg_file:
    cfg = json.load(cli_cfg_file)

if cfg['id'] != "dpm":
    raise ValueError('Configuration id mismatch: {}'.format(cfg['id']))

psi0 = cfg['psi0']
e0 = energy(psi0[0], psi0[1])

with open('conf/dpm/plot-orbit.yaml', 'r') as plot_opt_file:
    opt = yaml.load(plot_opt_file, Loader=yaml.Loader)

# Import calculated points
data = np.genfromtxt(
    os.path.join(opt['src']['dir'], opt['src']['data']),
    delimiter='\t',
    skip_header=1
)

# Plot orbit
def mk_orbit_plot(x, y):
    fig = plt.figure(figsize=(6, 8))
    plt.title("Double pendulum")
    plt.axis("equal")
    plt.xlabel("$\\theta_1$")
    plt.ylabel("$p_1$")
    plt.grid(True)
    plt.plot(
        x, y,
        fillstyle='none',
        marker='o',
        markersize=1.0,
        linestyle='None',
        alpha=0.25
    )
    xmin, xmax, ymin, ymax = plt.axis()
    plt.text(
        xmin - 0.15*(xmax - xmin),
        ymax - 0.05*(ymax - ymin),
        "$E$ = {:.4f}".format(e0)
    )
    return fig

def mk_inv_plot(x, y):
    n_start = divmod(2*len(x), 3)[0]
    n_stop = divmod(17*len(x), 24)[0]
    fig = plt.figure(figsize=(8, 8))
    plt.subplot(2, 1, 1)
    plt.title("Double pendulum. Invariant error")
    plt.xlabel("$t$")
    plt.ylabel("$\\delta E$")
    plt.grid(True)
    plt.plot(
        x, y,
        alpha=0.8
    )
    plt.subplot(2, 1, 2)
    plt.xlabel("$t$")
    plt.ylabel("$\\delta E$")
    plt.grid(True)
    plt.plot(
        x[n_start:n_stop], y[n_start:n_stop],
        alpha=0.8
    )
    return fig

# Plot orbit
fig = mk_orbit_plot(data[:, 1], data[:, 2])
plt.show()
fig.savefig(os.path.join(opt['dst']['dir'], opt['dst']['plot']['orbit']['q1p1']), dpi=150)

# Plot invariant error
fig = mk_inv_plot(data[:, 0], data[:, 3])
plt.show()
fig.savefig(os.path.join(opt['dst']['dir'], opt['dst']['plot']['invariant']['error']), dpi=150)

File: euler-ode/octave/dpm/plot_orbit.m
% Create plots using prepared tab separated values

% Config
cfg = jsondecode(fileread('../../conf/config.json'));
if(!strcmp(cfg.id, "dpm"))
  beep();
  error(["Configuration id mismatch: ", cfg.id]);
endif

opt = jsondecode(fileread('plot_orbit.json'));

% Problem invariant
function e = energy(q, p)
  cdiff = cos(q(1) - q(2));
  e = (p(1)^2 + 2*p(2)^2 - 2*p(1)*p(2)*cdiff)/(2*(2 - cdiff^2)) + 3 - 2*cos(q(1)) - cos(q(2));
endfunction

global E0;
E0 = energy(cfg.psi0(1, :), cfg.psi0(2, :));

% Import calculated points
data = dlmread(fullfile(opt.src.dir, opt.src.data), "\t", 1, 0);

% Plot orbit
function hf = orbit_plot(x, y)

  global E0;
  marker_size = 0.25;
  gap = 0.05;

  p_max = 2*sqrt(E0);
  q_max = acos(1 - E0/2);

  hf = figure();
  hold on
  grid on
  axis equal
  axis([-q_max - gap, q_max + gap, -p_max - gap, p_max + gap]);
  xlabel "\\theta_1"
  ylabel "p_1"
  set(gca, 'fontsize', 12);
  title 'Double pendulum'

  scatter(x, y, marker_size, 'r');

  glim = axis();
  text(
    glim(1) + 0.05*(glim(2) - glim(1)),
    glim(4) - 0.05*(glim(4) - glim(3)),
    sprintf("E = %d", E0)
  );
endfunction

% Plot invariant error
function hf = inv_plot(x, y)
  n_start = idivide(2*size(x, 1), int32(3), "fix");
  n_stop = idivide(17*size(x, 1), int32(24), "fix");
  hf = figure();
  hold on
  subplot(2, 1, 1);
  plot(x, y);
  xlabel("t")
  ylabel("\\delta E")
  grid on
  title 'Double pendulum. Invariant error'
  subplot(2, 1, 2);
  plot(x(n_start:n_stop), y(n_start:n_stop));
  xlabel("t")
  ylabel("\\delta E")
  grid on
  title 'Double pendulum. Invariant error (zoomed)'
endfunction

% Plot orbit
hf = orbit_plot(data(:, [2]), data(:, [3]));
print(hf, fullfile(opt.dst.dir, opt.dst.plot.q1p1), "-S720,960");

% Plot invariant error
hf = inv_plot(data(:, 1), data(:, 4));
print(hf, fullfile(opt.dst.dir, opt.dst.inv.error), "-S960,720");

